packages/world/src/World.sol

```diff
@@ -0,0 +1,422 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { console } from "forge-std/console.sol";
+import { Store, IStoreHook } from "store/Store.sol";
+import { StoreCore } from "store/StoreCore.sol";
+import { Schema } from "store/Schema.sol";
+
+import { OwnerTable, tableId as OwnerTableId } from "./tables/OwnerTable.sol";
+import { RouteAccessTable } from "./tables/RouteAccessTable.sol";
+import { RouteTable } from "./tables/RouteTable.sol";
+import { SystemRouteTable } from "./tables/SystemRouteTable.sol";
+import { SystemTable } from "./tables/SystemTable.sol";
+import { System } from "./System.sol";
+import { ISystemHook } from "./ISystemHook.sol";
+
+bytes32 constant ROOT_ROUTE_ID = keccak256(bytes(""));
+bytes32 constant SINGLE_SLASH = "/";
+
+contract World is Store {
+  error RouteInvalid(string route);
+  error RouteExists(string route);
+  error RouteAccessDenied(string route, address caller);
+  error SystemExists(address system);
+
+  constructor() {
+    SystemTable.registerSchema();
+    RouteTable.registerSchema();
+    RouteAccessTable.registerSchema();
+    SystemRouteTable.registerSchema();
+    OwnerTable.registerSchema();
+
+    // Register root route and give ownership to msg.sender
+    RouteTable.set({ routeId: ROOT_ROUTE_ID, route: "" }); // Storing this explicitly to trigger the event for indexers
+    OwnerTable.set({ key: ROOT_ROUTE_ID, owner: msg.sender });
+    RouteAccessTable.set({ routeId: ROOT_ROUTE_ID, caller: msg.sender, access: true });
+  }
+
+  /************************************************************************
+   *
+   *    REGISTRATION METHODS
+   *
+   ************************************************************************/
+
+  /**
+   * Register a new route by extending an existing route
+   */
+  function registerRoute(string calldata baseRoute, string calldata subRoute) public returns (bytes32 routeId) {
+    // Require subroute to be a valid route fragment (start with `/` and don't contain any further `/`)
+    if (!_isSingleLevelRoute(subRoute)) revert RouteInvalid(subRoute);
+
+    // Require base route to exist (with a special check for the root route because it's empty and fails the `has` check)
+    if (!(keccak256(bytes(baseRoute)) == ROOT_ROUTE_ID || RouteTable.has(keccak256(bytes(baseRoute)))))
+      revert RouteInvalid(baseRoute);
```

Comment:
```suggestion

    // Require base route to exist (with a special check for the root route because it's empty and fails the `has` check)
    bytes32 baseRouteId = keccak256(bytes(baseRoute));
    if (!(baseRouteId == ROOT_ROUTE_ID || RouteTable.has(baseRouteId)))
      revert RouteInvalid(baseRoute);
```